File: backend/LessonRunner.Domain/Lessons/Lesson.cs
using LessonRunner.Domain.Common;

namespace LessonRunner.Domain.Lessons;

public sealed class Lesson : Entity
{
    /// <summary>Format zajęć. Brak pola w starszym dokumencie JSON daje wartość Standard.</summary>
    public LessonKind Kind { get; set; } = LessonKind.Standard;
    public required string Title { get; set; }
    public required string Subject { get; set; }
    public required string Level { get; set; }
    public required string Description { get; set; }
    public int Order { get; set; }
    public LessonStatus Status { get; set; } = LessonStatus.Draft;
    public DateTimeOffset CreatedAt { get; init; } = DateTimeOffset.UtcNow;
    public DateTimeOffset UpdatedAt { get; set; } = DateTimeOffset.UtcNow;
    public List<string> Tags { get; set; } = [];
    public List<LessonStep> Steps { get; set; } = [];
    public LessonProjectFiles ProjectFiles { get; set; } = new();

    /// <summary>
    /// Po co są te zajęcia — jedno zdanie dla prowadzącego.
    ///
    /// `Description` opisuje lekcję na zewnątrz (biblioteka, oferta). To jest cel dydaktyczny:
    /// czego dziecko ma się nauczyć. Do tej pory nie było go gdzie zapisać, więc prowadzący
    /// dostawał scenariusz bez odpowiedzi, co jest w nim najważniejsze.
    /// </summary>
    public string? Objective { get; set; }

    /// <summary>Po zajęciach dziecko potrafi… — sprawdzalne kryteria, nie życzenia.
    /// Są jednocześnie gotowym szkieletem notatki dla rodzica.</summary>
    public List<string> SuccessCriteria { get; set; } = [];

    /// <summary>Co przygotować przed zajęciami: plik startowy, konto, sprawdzony dźwięk.
    /// Rzeczy, które odkryte w trakcie lekcji kosztują pięć minut ciszy.</summary>
    public List<string> Preparation { get; set; } = [];

    /// <summary>Zadanie domowe / co pokazać rodzicom po zajęciach.</summary>
    public List<string> Homework { get; set; } = [];
}

File: frontend/lesson-runner-web/src/types/lesson.ts
export type LessonStatus = "draft" | "review" | "ready";
export type LessonKind = "standard" | "showcase";

export type LessonSummary = {
  id: string;
  title: string;
  subject: string;
  level: string;
  description: string;
  order: number;
  status: LessonStatus;
  statusLabel: string;
  stepCount: number;
  durationMinutes: number;
  kind: LessonKind;
  kindLabel: string;
  scheduledDurationMinutes: number;
  earlyLeaveAfterMinutes: number | null;
};

export type LessonDetails = LessonSummary & {
  tags: string[];
  projectFiles: LessonProjectFiles;
  steps: LessonStep[];
} & LessonIntent;

/**
 * Po co są te zajęcia i co po nich zostaje.
 *
 * `description` opisuje lekcję na zewnątrz (biblioteka), `objective` odpowiada prowadzącemu,
 * co jest w niej najważniejsze. Do lipca 2026 nie było gdzie tego zapisać.
 */
export type LessonIntent = {
  objective?: string | null;
  /** Po zajęciach dziecko potrafi… */
  successCriteria?: string[] | null;
  /** Co przygotować przed zajęciami. */
  preparation?: string[] | null;
  /** Zadanie domowe albo co pokazać rodzicom. */
  homework?: string[] | null;
};

export type LessonProjectFiles = {
  starter?: LessonProjectFile | null;
  final?: LessonProjectFile | null;
};

export type LessonProjectFile = {
  label: string;
  url: string;
  fileName: string;
  contentType: string;
  sizeBytes: number;
  downloadToken?: string | null;
  downloadUrl?: string | null;
};

export type LessonStep = {
  id: string;
  order: number;
  type: string;
  title: string;
  durationMinutes: number;
  script: string[];
  studentItems: StudentItem[];
  resources: LessonResource[];
  notes: LessonNote[];
};

export type StudentItem = {
  kind: string;
  text?: string | null;
  caption?: string | null;
  url?: string | null;
};

export type LessonResource = {
  kind: string;
  label?: string | null;
  url?: string | null;
  code?: string | null;
  language?: string | null;
};

export type LessonNote = {
  kind: string;
  text: string;
};

export type StoredFile = {
  url: string;
  fileName: string;
  contentType: string;
  sizeBytes: number;
};

export type LessonRunSession = {
  id: string;
  lessonId: string;
  userId: string;
  stepIndex: number;
  elapsedTotalSeconds: number;
  elapsedStepSeconds: number;
  running: boolean;
  startedAt: string;
  updatedAt: string;
};

export type UpdateLessonRunSessionRequest = {
  stepIndex: number;
  elapsedTotalSeconds: number;
  elapsedStepSeconds: number;
  running: boolean;
};

export type CreateLessonRequest = {
  kind: LessonKind;
  title: string;
  subject: string;
  level: string;
  description: string;
  tags: string[];
  projectFiles?: LessonProjectFiles | null;
  steps: CreateLessonStepRequest[];
  order?: number;
} & LessonIntent;

export type CreateLessonStepRequest = {
  type: string;
  title: string;
  durationMinutes: number;
  script: string[];
  studentItems: StudentItem[];
  resources: LessonResource[];
  notes: LessonNote[];
};
